src/ApiClient.js

```diff
@@ -0,0 +1,71 @@
+// This class is a reference implementation. If you desire additional
+// functionality, you can work with the config options here, or copy this code
+// into your project and customize the ApiClient to your needs.
+
+class ApiClient {
+  constructor(passedConfig) {
+    const defaultConfig = {
+      methods: ['get', 'post', 'put', 'patch', 'delete'],
+      credentials: 'same-origin',
+      defaultHeaders: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json'
+      },
+      format: 'json' // json, blob, text, arrayBuffer, ...
+    }
+
+    const config = {
+      ...defaultConfig,
+      ...passedConfig
+    }
+
+    if (!config.basePath) {
+      // e.g. 'https://example.com/api/v3'
+      throw new Error('You must pass a base path to the ApiClient')
+    }
+
+    config.methods.forEach(method => {
+      this[method] = (path, { params, data, otherConfig } = {}) => {
+        const queryString = this.queryString(params)
+        const { basePath, defaultHeaders, format, ...fetchConfig } = config
+        const headers = {
+          this.getHeaders(defaultHeaders, data),
+          ...fetchConfig.headers
+        }
+
+        return new Promise((resolve, reject) => {
+          fetch(basePath + path + this.queryString(), {
+            ...fetchConfig,
+            ...otherConfig,
+            method,
+            headers
+          }).then(response => {
+            return response[format]()
+          }).then(payload => {
+            resolve(payload)
+          }).catch(error => {
+            reject(error)
+          })
```

Comment:
Unless there is something special about implementation of `Promise`, it is not necessary to wrap this result in a promise constructor. `fetch` returns a promise by itself. This would work as well:

```js
return fetch(basePath + path + this.queryString(), {
  ...fetchConfig,
  ...otherConfig,
  method,
  headers
}).then(response => {
  return response[format]()
})
```